File: cli/internal/agentauthz/auth_blob.go
package agentauthz

import (
	"fmt"
	"io"
	"os"
	"path/filepath"
	"runtime"
	"strings"

	"harness-data/cli/internal/harness"
)

const (
	EnvAuthBlob            = "HARNESS_AUTH_BLOB"
	EnvAuthBlobFile        = "HARNESS_AUTH_BLOB_FILE"
	EnvAuthUserID          = "HARNESS_AUTH_USER_ID"
	EnvRequesterContextDir = "LUMI_REQUESTER_CONTEXT_DIR"
	EncryptedBlobPrefix    = "qdm1enc."
	MaxCredentialBytes     = 1 << 20
)

type BlobSource string

const (
	BlobSourceEnv     BlobSource = "env"
	BlobSourceEnvFile BlobSource = "env_file"
	BlobSourceFile    BlobSource = "file"
)

type ResolvedBlob struct {
	Blob string
	// UserID is retained for compatibility with callers that inspect the
	// resolved source. It is never passed to qdm or used for authorization.
	UserID string
	Source BlobSource
}

type ResolveOptions struct {
	ProjectRoot string
	Config      harness.AuthzConfig
	Env         map[string]string
}

func ResolveAuthBlob(opts ResolveOptions) (ResolvedBlob, error) {
	env := opts.Env
	if env == nil {
		env = environMap()
	}

	if !opts.Config.LocalBlobAllowed() {
		return ResolvedBlob{}, fmt.Errorf("authz mode is on but local blob fallback is disabled (allow_local_blob=false)")
	}

	if blob := strings.TrimSpace(env[EnvAuthBlob]); blob != "" {
		if !IsEncryptedBlob(blob) {
			return ResolvedBlob{}, fmt.Errorf("HARNESS_AUTH_BLOB must be an encrypted qdm1enc blob")
		}
		return ResolvedBlob{Blob: blob, UserID: localUserID(env, opts.Config), Source: BlobSourceEnv}, nil
	}

	if path := strings.TrimSpace(env[EnvAuthBlobFile]); path != "" {
		blob, err := readBlobFile(opts.ProjectRoot, path)
		if err != nil {
			return ResolvedBlob{}, err
		}
		return ResolvedBlob{Blob: blob, UserID: localUserID(env, opts.Config), Source: BlobSourceEnvFile}, nil
	}

	if path := strings.TrimSpace(opts.Config.BlobFile); path != "" {
		blob, err := readBlobFile(opts.ProjectRoot, path)
		if err != nil {
			return ResolvedBlob{}, err
		}
		userID := strings.TrimSpace(opts.Config.DevUserID)
		return ResolvedBlob{Blob: blob, UserID: userID, Source: BlobSourceFile}, nil
	}

	return ResolvedBlob{}, fmt.Errorf("authz mode is on but no encrypted blob is available (HARNESS_AUTH_BLOB, HARNESS_AUTH_BLOB_FILE, or authz.blob_file)")
}

func IsEncryptedBlob(value string) bool {
	return strings.HasPrefix(strings.TrimSpace(value), EncryptedBlobPrefix)
}

func localUserID(env map[string]string, cfg harness.AuthzConfig) string {
	if userID := strings.TrimSpace(env[EnvAuthUserID]); userID != "" {
		return userID
	}
	return strings.TrimSpace(cfg.DevUserID)
}

func readBlobFile(projectRoot, pathValue string) (string, error) {
	pathValue = strings.TrimSpace(pathValue)
	if pathValue == "" {
		return "", fmt.Errorf("auth blob file path is empty")
	}
	absolute := pathValue
	if !filepath.IsAbs(absolute) {
		absolute = filepath.Join(projectRoot, filepath.FromSlash(pathValue))
	}
	file, err := openCredentialFile(absolute)
	if err != nil {
		return "", fmt.Errorf("auth blob file unavailable: %s", absolute)
	}
	defer file.Close()
	before, err := file.Stat()
	if err != nil || !before.Mode().IsRegular() {
		return "", fmt.Errorf("auth blob file must be a regular file: %s", absolute)
	}
	if runtime.GOOS != "windows" && before.Mode().Perm()&0o077 != 0 {
		return "", fmt.Errorf("auth blob file permissions must be 0600: %s", absolute)
	}
	if before.Size() > MaxCredentialBytes {
		return "", fmt.Errorf("auth blob file exceeds maximum size: %s", absolute)
	}
	data, err := io.ReadAll(io.LimitReader(file, MaxCredentialBytes+1))
	if err != nil {
		return "", fmt.Errorf("auth blob file unavailable: %s", absolute)
	}
	if len(data) > MaxCredentialBytes {
		return "", fmt.Errorf("auth blob file exceeds maximum size: %s", absolute)
	}
	after, err := file.Stat()
	if err != nil || !os.SameFile(before, after) || before.Size() != after.Size() || before.Mode() != after.Mode() || !before.ModTime().Equal(after.ModTime()) {
		return "", fmt.Errorf("auth blob file changed while reading: %s", absolute)
	}
	blob := strings.TrimSpace(string(data))
	if blob == "" {
		return "", fmt.Errorf("auth blob file is empty: %s", absolute)
	}
	if !IsEncryptedBlob(blob) {
		return "", fmt.Errorf("auth blob file must contain a qdm1enc blob: %s", absolute)
	}
	return blob, nil
}

func environMap() map[string]string {
	out := map[string]string{}
	for _, item := range os.Environ() {
		key, value, ok := strings.Cut(item, "=")
		if ok {
			out[key] = value
		}
	}
	return out
}

File: packages/data-harness-cli/src/lib/authz/auth-blob.js
import { lstatSync, readFileSync, realpathSync } from "node:fs";
import path from "node:path";

import { localBlobAllowed } from "../harness.js";
import {
  ENCRYPTED_BLOB_PREFIX,
  ENV_AUTH_BLOB,
  ENV_AUTH_BLOB_FILE,
  ENV_AUTH_USER_ID,
} from "./constants.js";
import { environMap } from "./env.js";

export const BLOB_SOURCE_ENV = "env";
export const BLOB_SOURCE_ENV_FILE = "env_file";
export const BLOB_SOURCE_FILE = "file";
export const BLOB_SOURCE_SECRET_REF_FILE = "secret_ref_file";

export function isEncryptedBlob(value) {
  return String(value || "").trim().startsWith(ENCRYPTED_BLOB_PREFIX);
}

/**
 * @param {{ projectRoot: string, config: import("../harness.js").AuthzConfig, env?: Record<string, string>, secretRef?: { kind: string, path?: string } | null }} opts
 */
export function resolveAuthBlob(opts) {
  const env = opts.env || environMap();
  if (opts.secretRef) {
    if (opts.secretRef.kind !== "file") {
      throw new Error(`secret reference kind is not supported by the local adapter: ${opts.secretRef.kind}`);
    }
    const file = readBlobFileAbsolute(opts.secretRef.path);
    const userId = localUserID(env, opts.config);
    if (!userId) {
      throw new Error("secretRef file requires HARNESS_AUTH_USER_ID or authz.dev_user_id");
    }
    return { blob: file.blob, sourcePath: file.path, userId, source: BLOB_SOURCE_SECRET_REF_FILE };
  }

  if (!localBlobAllowed(opts.config)) {
    throw new Error("authz mode is on but local blob fallback is disabled (allow_local_blob=false)");
  }

  const envBlob = String(env[ENV_AUTH_BLOB] || "").trim();
  if (envBlob) {
    if (!isEncryptedBlob(envBlob)) {
      throw new Error("HARNESS_AUTH_BLOB must be an encrypted qdm1enc blob");
    }
    const userId = localUserID(env, opts.config);
    if (!userId) {
      throw new Error("authz local blob requires HARNESS_AUTH_USER_ID or authz.dev_user_id");
    }
    return { blob: envBlob, userId, source: BLOB_SOURCE_ENV };
  }

  const envFile = String(env[ENV_AUTH_BLOB_FILE] || "").trim();
  if (envFile) {
    const file = readBlobFile(opts.projectRoot, envFile);
    const userId = localUserID(env, opts.config);
    if (!userId) {
      throw new Error("authz local blob requires HARNESS_AUTH_USER_ID or authz.dev_user_id");
    }
    return { blob: file.blob, sourcePath: file.path, userId, source: BLOB_SOURCE_ENV_FILE };
  }

  const configFile = String(opts.config?.blobFile || "").trim();
  if (configFile) {
    const file = readBlobFile(opts.projectRoot, configFile);
    const userId = String(opts.config?.devUserId || "").trim();
    if (!userId) {
      throw new Error("authz.blob_file requires authz.dev_user_id (no default principal)");
    }
    return { blob: file.blob, sourcePath: file.path, userId, source: BLOB_SOURCE_FILE };
  }

  throw new Error(
    "authz mode is on but no encrypted blob is available (HARNESS_AUTH_BLOB, HARNESS_AUTH_BLOB_FILE, or authz.blob_file)",
  );
}

function localUserID(env, cfg) {
  const fromEnv = String(env[ENV_AUTH_USER_ID] || "").trim();
  if (fromEnv) return fromEnv;
  return String(cfg?.devUserId || "").trim();
}

function readBlobFile(projectRoot, pathValue) {
  pathValue = String(pathValue || "").trim();
  if (!pathValue) throw new Error("auth blob file path is empty");
  const absolutePath = path.isAbsolute(pathValue) ? pathValue : path.join(projectRoot, pathValue);
  return readBlobFileAbsolute(absolutePath);
}

function readBlobFileAbsolute(absolute) {
  let info;
  try {
    info = lstatSync(absolute);
  } catch {
    throw new Error(`auth blob file not found: ${absolute}`);
  }
  if (info.isSymbolicLink() || !info.isFile()) {
    throw new Error(`auth blob file must be a regular file: ${absolute}`);
  }
  if (process.platform !== "win32" && (info.mode & 0o777) !== 0o600) {
    throw new Error(`auth blob file permissions must be 0600: ${absolute}`);
  }
  const blob = readFileSync(absolute, "utf8").trim();
  if (!blob) throw new Error(`auth blob file is empty: ${absolute}`);
  if (!isEncryptedBlob(blob)) {
    throw new Error(`auth blob file must contain a qdm1enc blob: ${absolute}`);
  }
  return { blob, path: realpathSync(absolute) };
}
